### DISCUSSION TITLE
out of memory...again

### DISCUSSION THREAD

**Discussion by CarlFredrikFallen** [talmolab/sleap#1592] [Help!]
First of all, many thanks for a great tool! Everything worked fine, but then I upgraded windows, and of course the GPU couldnt be found again. I reinstalled everything from scratch, GPU is found, but when I run training, it seems the GPU is not initiated. And it complains that it has ran out of memory. Noticed there was someone with similar issue, and suggestion was just to restart computer...this I have done multiple times now. What else could be wrong here? Many thanks! 

[sleapoutput.txt](https://github.com/talmolab/sleap/files/13327948/sleapoutput.txt)

**Comment by talmo** [comment]
Hi @CarlFredrikFallen,

Apologies for the delay!

I'm pulling out the relevant bits from your log file (thanks for including this!!):

```
INFO:sleap.nn.training:Versions:
SLEAP: 1.3.3
TensorFlow: 2.7.0
Numpy: 1.21.6
Python: 3.7.12
OS: Windows-10-10.0.22621-SP0
...
        "preprocessing": {
            "ensure_rgb": false,
            "ensure_grayscale": false,
            "imagenet_mode": null,
            "input_scaling": 0.5,
            "pad_to_stride": 16,
            "resize_and_pad_to_target": true,
            "target_height": 1560,
            "target_width": 1560
        },
...
    "model": {
        "backbone": {
            "leap": null,
            "unet": {
                "stem_stride": null,
                "max_stride": 16,
                "output_stride": 2,
                "filters": 16,
                "filters_rate": 2.0,
                "middle_block": true,
                "up_interpolate": true,
                "stacks": 1
            },
            "hourglass": null,
            "resnet": null,
            "pretrained_encoder": null
        },
        "heads": {
            "single_instance": null,
            "centroid": {
                "anchor_part": "thorax",
                "sigma": 2.5,
                "output_stride": 2,
                "loss_weight": 1.0,
                "offset_refinement": false
            },
...
INFO:sleap.nn.training:Auto-selected GPU 0 with 12011 MiB of free memory.
INFO:sleap.nn.training:Using GPU 0 for acceleration.
INFO:sleap.nn.training:Disabled GPU memory pre-allocation.
INFO:sleap.nn.training:System:
GPUs: 1/1 available
  Device: /physical_device:GPU:0
         Available: True
        Initalized: False
     Memory growth: True
...
2023-11-12 12:50:48.135387: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 9399 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4080 Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.9
...
INFO:sleap.nn.training:  Input shape: (784, 784, 1)
INFO:sleap.nn.training:Created Keras model.
INFO:sleap.nn.training:  Backbone: UNet(stacks=1, filters=16, filters_rate=2.0, kernel_size=3, stem_kernel_size=7, convs_per_block=2, stem_blocks=0, down_blocks=4, middle_block=True, up_blocks=3, up_interpolate=True, block_contraction=False)
INFO:sleap.nn.training:  Max stride: 16
INFO:sleap.nn.training:  Parameters: 1,953,105
INFO:sleap.nn.training:  Heads:
INFO:sleap.nn.training:    [0] = CentroidConfmapsHead(anchor_part='thorax', sigma=2.5, output_stride=2, loss_weight=1.0)
INFO:sleap.nn.training:  Outputs:
INFO:sleap.nn.training:    [0] = KerasTensor(type_spec=TensorSpec(shape=(None, 392, 392, 1), dtype=tf.float32, name=None), name='CentroidConfmapsHead/BiasAdd:0', description="created by layer 'CentroidConfmapsHead'")
...
2023-11-12 12:52:17.917851: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.65GiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.
```

Ok so, let's break this down:

> ```
> INFO:sleap.nn.training:System:
> GPUs: 1/1 available
>   Device: /physical_device:GPU:0
>          Available: True
>         Initalized: False
>      Memory growth: True
> ```

This totally fine -- this is the state that the GPU starts at before SLEAP actually loads the model up on the GPU.


> ```
> INFO:sleap.nn.training:Auto-selected GPU 0 with 12011 MiB of free memory.
> ...
> 2023-11-12 12:50:48.135387: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 9399 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4080 Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.9
> ```

As far as GPU memory goes, you're well above the recommended capacity with ~12 GB of GPU memory on a GeForce RTX 4080 (laptop edition). I'm not sure why it reports it as only ~9 GB later on, but either way this is still plenty.

> ```
> INFO:sleap.nn.training:  Input shape: (784, 784, 1)
> ...
> INFO:sleap.nn.training:  Parameters: 1,953,105
> INFO:sleap.nn.training:  Heads:
> INFO:sleap.nn.training:    [0] = CentroidConfmapsHead(anchor_part='thorax', sigma=2.5, output_stride=2, loss_weight=1.0)
> INFO:sleap.nn.training:  Outputs:
> INFO:sleap.nn.training:    [0] = KerasTensor(type_spec=TensorSpec(shape=(None, 392, 392, 1), dtype=tf.float32, name=None), name='CentroidConfmapsHead/BiasAdd:0', description="created by layer 'CentroidConfmapsHead'")
> ```

The model itself is fairly small, with only about <2M parameters. The input layer size is kind of large, which means big tensors will be initialized in the first few layers, but it's nothing crazy and something that your GPU should have plenty of capacity to handle. You'll only start running into problems at ~1024 x 1024 or larger I would estimate.

> ```
> 2023-11-12 12:52:17.917851: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.65GiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.
> ```

I'm guessing this is the message that prompted your post. This is actually a warning and totally safe to ignore. Note the `W` in front and the clarification that `this is not a failure, but may mean that there could be performance gains if more memory were available`.

If you're not seeing a crash subsequent to this message, you should be good to go. If you're actually having memory issues, then you'd see a big spam in the terminal with lots of memory addresses, or something like this:

```
E tensorflow/compiler/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:306] gpu_async_0 cuMemAllocAsync failed to allocate 957465616 bytes: CUDA error: out of memory (CUDA_ERROR_OUT_OF_MEMORY)
```

The reason you get those warnings is because SLEAP (via TensorFlow) tries to optimize memory usage for maximum performance, and in doing so uses some heuristics to figure out the best memory allocation. Sometimes it doesn't get it quite right so it throws that warning that it tried to optimize further, but couldn't because it ran out of memory for a given operation. When that happens it just falls back to a slower version of the operation that is less memory intensive.

If you're running into errors like described above or if SLEAP is crashing let us know and we can troubleshoot from there.

Cheers,

Talmo
